csrc/scheduler/matmul.cpp

```diff
@@ -711,6 +711,17 @@ void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
       mma_ops.size() == 1,
       "scheduleMatmul supports fusion with single mma op in definition, got ",
       mma_ops.size());
+
+  // Skip scheduling if Matmul will be expression evaluated.
+  if (isOptionEnabled(EnableOption::MatmulExprEval)) {
```

Comment:
Does this mean any matmul pattern including mma + epilogue will be handled by the expression evaluator? Shouldn't it only take care of the mma part?

Reply:
There are two problems for taking care of just the mma: 
1. at::matmul doesn't do HH->S. We could plug in another backend that supports HH->S for EE though. 
2. MMA is never alone in these GPT models (e.g. LLaMA). It's always part of a linear layer or an SDPA. nvFuser doesn't do SDPA well and we will have to offload it to another executor for quite some time, so scratch that. A linear layer however comes with this MMA->BiasAdd->Relu pattern. In order for its performance to be on par with framework-not-giving-nvFuser-the-matmul, we have to execute MMA+epilogue in one call. 

Wdyt, @naoyam? Extending EE to do MMA+epilogue is the most obvious way to me to solve the above problems. But I could definitely be wrong.

Reply:
> Extending EE to do MMA+epilogue

How would it do that? Does Aten support matmul with some epilogue op?

Reply:
Oh yeah, it's called `at::addmm`. I realized I was wrong about Relu. [torch.nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) doesn't do Relu, so the pattern would be matmul+biasadd which is what `at::addmm` does.

Reply:
So, how would you handle matmul+epilogue patterns that are accepted by the nvFuser matmul scheduler but there's no corresponding aten version? Would that end up doing mma and the epilogue op separately?

Reply:
> Would that end up doing mma and the epilogue op separately?

Yes. However, we should probably change MatmulScheduler::canSchedule to avoid pulling in any non-fusible epilogue in the first place (so it has a chance to be fused with the following segment). Although the exact heuristics are unclear to me, I imagine MatmulScheduler::canSchedule will first decide whether to do matmul through codegen or EE, and then pull in as many epilogue ops as **that** backend (i.e. codegen or EE) can fuse.

Reply:
That makes sense. Or, maybe we might want to have a separate scheduler for EE? We try the native matmul scheduler first, and then the EE matmul scheduler, and then the other schedulers. No particular preference but just my two cents.

Reply:
Yes, that's a certainly great option to consider. It makes things more composable at a risk of being harder to share logics with MatmulScheduler. I hope the preference will be more obvious when we know what the heuristics look like!

Reply:
Adding on to @wujingyue's comments, the plan for the next PRs is:

1. Support Mma + Cast -> avoid roundtrip casting (`half-> float -> half`) by checking `mma_out->uses()` and if it is either a castOp or pointwise ops with inputs of the same type (`half`), then skip casting back the output of `at::matmul`. This will not execute `matmul + bias` in a single call.
2. Handle common epilogue fusions: We will need to pattern match and evaluate within the MmaOp accordingly. `test_matmul_scheduler.cpp` currently has a few cases that I will start with: `mma + bias`, `mma + bias + relu/gelu`, `mma + relu`.
3. Epilogue fusions that are not be supported: They can still be computed through EE but should ideally not be plumbed down to the matmul scheduler.

Reply:
Do 2 only when it's needed. You should double check this but I think Llama for example uses linear with bias off and none of the linear layers in our benchmarks do relu or gelu

Reply:
Llama2 has `bias=False` in the linear layers but some GPT configs have `bias=True`.